design-proposals/klip-38-variable-substitution.md

```diff
@@ -264,7 +264,21 @@ Scope: SESSION
 In Java API client, the session is opened when `Client.create()` is called, and closed when `Client.close()` is called. Variables
 may be defined using a client method instead of a new syntax. This is to ensure users know the scope starts after the `create()` call.
 
-TBD: Define a Java API method for variable substitution.
+Users will define variables using the functions, `define(variable_name, value)` and `undefine(variable_name)`.
+The client will keep track of the calls made in a map. Alternatively, the `executeStatement` and `executeQuery`
+functions could accept a third parameter, `sessionVariables` that is a map of variables to variable definitions.
+
+There are two options for executing the variable execution. The first is to do it in the client by
+using the variable substitutor in `ksqldb-parser`. The disadvantages to this are that `ksqldb-parser`
```

Comment:
I haven't looked into the code for this at all but is it not possible to separate out parser functionality for variable substitution into a library in ksqldb-common? Or is the variable substitution somehow intertwined with the actual statement parsing itself.

Reply:
The statements have to be parsed because the substitutor needs to be able to differentiate between literals, identifiers and keywords (substitution is not allowed for keywords).

Reply:
Ah bummer... in light of that, if we think this is the least of all evils and @colinhicks is OK with the added complexity to our HTTP endpoints, then I'm happy with this approach.